aptos-move/framework/aptos-framework/sources/PropertyList.move

```diff
@@ -0,0 +1,182 @@
+/// This Data Structure is designed to hold properties.
+/// It only support
+///     1. adding new property.
+///     2. update the value of existing property
+///     3. get the value of a property
+module AptosFramework::PropertyList {
+    use Std::Vector;
+    use Std::ASCII::String;
+    use AptosFramework::SimpleMap::{Self, SimpleMap};
+
+    const MAX_PROPERTY_MAP_SIZE: u64 = 1000;
+    const EKEY_AREADY_EXIST_IN_PROPERTY_MAP: u64 = 1;
+    const EPROPERTY_NUMBER_EXCEED_LIMIT: u64 = 2;
+    const EPROPERTY_NOT_EXIST: u64 = 3;
+    const EKEY_COUNT_NOT_MATCH_VALUE_COUNT: u64 = 4;
+    const EKEY_COUNT_NOT_MATCH_TYPE_COUNT: u64 = 5;
+
+    struct PropertyList has store {
+        map: SimpleMap<String, PropertyValue>
+    }
+
+    struct PropertyValue has store, copy, drop {
+        value: vector<u8>,
+        type: String,
+    }
+
+    public fun new(
+        default_keys: &vector<String>,
+        default_values: &vector<vector<u8>>,
+        default_types: &vector<String>
+    ): PropertyList {
+        assert!(Vector::length(default_keys) == Vector::length(default_values), EKEY_COUNT_NOT_MATCH_VALUE_COUNT);
+        assert!(Vector::length(default_keys) == Vector::length(default_types), EKEY_COUNT_NOT_MATCH_TYPE_COUNT);
+        let properties = PropertyList{
+            map: SimpleMap::create<String, PropertyValue>()
+        };
+        let i = 0;
+        while (i < Vector::length(default_keys)) {
+            SimpleMap::add(
+                &mut properties.map,
+                *Vector::borrow(default_keys, i),
+                PropertyValue{ value: *Vector::borrow(default_values, i), type: *Vector::borrow(default_types, i) }
+            );
+            i = i + 1;
+        };
+        properties
+    }
+
+    public fun contains(list: &PropertyList, key: &String): bool {
+        SimpleMap::contains_key(&list.map, key)
+    }
+
+    public fun add(list: &mut PropertyList, key: String, value: PropertyValue) {
+        assert!(! SimpleMap::contains_key(&list.map, &key), EKEY_AREADY_EXIST_IN_PROPERTY_MAP);
+        assert!(SimpleMap::length<String, PropertyValue>(&list.map) <= MAX_PROPERTY_MAP_SIZE, EPROPERTY_NUMBER_EXCEED_LIMIT);
+        SimpleMap::add(&mut list.map, key, value);
+    }
+
+    public fun size(list: &PropertyList): u64 {
+        SimpleMap::length(&list.map)
+    }
+
+    public fun borrow(list: &PropertyList, key: &String): &PropertyValue {
+        let found = contains(list, key);
+        assert!(found, EPROPERTY_NOT_EXIST);
+        SimpleMap::borrow(&list.map, key)
+    }
+
+    public fun get_value(list: &PropertyList, key: &String): vector<u8> {
+        let found = contains(list, key);
+        assert!(found, EPROPERTY_NOT_EXIST);
+        SimpleMap::borrow(&list.map, key).value
+    }
+
+    public fun get_type(list: &PropertyList, key: &String, ): String {
+        let found = contains(list, key);
+        assert!(found, EPROPERTY_NOT_EXIST);
+        SimpleMap::borrow(&list.map, key).type
+    }
```

Comment:
This doesn't feel ergonomic and I doubt we'll ever really need these, but I'd rather have helpers on the PropertyValue instead of on the PropertyMap. That is you use the borrow above and then you have borrow_value and borrow_type that takes in the PropertyValue.

Reply:
updated to new function as recommended